modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/LocalFileConfigurationStorage.java

```diff
@@ -0,0 +1,224 @@
+                    ConfigValue value = e.getValue();
+                    if (value.valueType() == ConfigValueType.LIST) {
+                        map.put(key, toArray(lastState.get(key).getClass(), (ConfigList) value));
+                    } else {
+                        map.put(key, (Serializable) value.unwrapped());
+                    }
+                }
+            }
+        }
+        return CompletableFuture.completedFuture(map);
+    }
+
+    @Override
+    public CompletableFuture<Serializable> readLatest(String key) {
+        ensureConfigFile();
+        return CompletableFuture.completedFuture((Serializable) parseConfigOptions().getValue(key).unwrapped());
+    }
+
+    @Override
+    public CompletableFuture<Boolean> write(Map<String, ? extends Serializable> newValues, long ver) {
+        ensureConfigFile();
+        saveValues(newValues);
+        lastState.putAll(newValues);
+        runAsync(() -> lsnrRef.get().onEntriesChanged(new Data(newValues, 0)));
+        return CompletableFuture.completedFuture(true);
+    }
+
+    private void runAsync(Runnable runnable) {
+        CompletableFuture<Void> future = CompletableFuture.runAsync(runnable, threadPool);
+
+        futureTracker.registerFuture(future);
+    }
+
+    @Override
+    public void registerConfigurationListener(ConfigurationStorageListener lsnr) {
+        if (!lsnrRef.compareAndSet(null, lsnr)) {
+            LOG.debug("Configuration listener has already been set");
+        }
+    }
+
+    @Override
+    public ConfigurationType type() {
+        return ConfigurationType.LOCAL;
+    }
+
+    @Override
+    public CompletableFuture<Long> lastRevision() {
+        return CompletableFuture.completedFuture(0L);
+    }
+
+    @Override
+    public CompletableFuture<Void> writeConfigurationRevision(long prevRevision, long currentRevision) {
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public void close() {
+        IgniteUtils.shutdownAndAwaitTermination(threadPool, 10, TimeUnit.SECONDS);
+
+        futureTracker.cancelInFlightFutures();
+    }
+
+    private String toString(Map<String, ? extends Serializable> values) {
+        Map<String, Object> map = values.entrySet().stream().collect(Collectors.toMap(Entry::getKey, stringEntry -> {
+            Serializable value = stringEntry.getValue();
+            if (value.getClass().isArray()) {
+                return Arrays.asList((Object[]) value);
+            }
+            return value;
+        }));
+        Config other = ConfigFactory.parseMap(map);
+        Config newConfig = parseConfigOptions().withFallback(other).resolve();
+        return newConfig.isEmpty()
+                ? ""
+                : newConfig.root().render(ConfigRenderOptions.defaults().setJson(false));
+    }
+
+    private void saveValues(Map<String, ? extends Serializable> values) {
+        try {
+            Files.write(configPath, toString(values).getBytes(StandardCharsets.UTF_8));
```

Comment:
Not safe file writing, once again. At first, you should write data in another file, and only then you have a right to corrupt/delete old file, because new one is already on disk.
On top of that, we need a "recovery" procedure that would delete tmp file or maybe rename it to proper config file if it's already missing